File: LeetCode547.cpp
#include<vector>
#include<iostream>
#include<queue>
using namespace std;

class Solution {
public:
    void dfs(vector<vector<int>>& isConnected, vector<int>& visited, int provinces, int i) {
        for (int j = 0; j < provinces; j++) {
            if (isConnected[i][j] == 1 && !visited[j]) {
                visited[j] = 1;
                dfs(isConnected, visited, provinces, j);
            }
        }
    }
    // 深度优先
    int findCircleNum1(vector<vector<int>>& isConnected) {
        int provinces = isConnected.size();
        vector<int> visited(provinces);
        int circles = 0;
        for (int i = 0; i < provinces; i++) {
            if (!visited[i]) {
                dfs(isConnected, visited, provinces, i);
                circles++;
            }
        }
        return circles;
    }
    // 广度优先
    int findCircleNum2(vector<vector<int>>& isConnected) {
        int provinces = isConnected.size();
        vector<int> visited(provinces);
        int circles = 0;
        queue<int> Q;
        for (int i = 0; i < provinces; i++) {
            if (!visited[i]) {
                Q.push(i);
                while (!Q.empty()) {
                    int j = Q.front();
                    Q.pop();
                    visited[j] = 1;
                    for (int k = 0; k < provinces; k++) {
                        if (isConnected[j][k] == 1 && !visited[k]) {
                            Q.push(k);
                        }
                    }
                }
                circles++;
            }
        }
        return circles;
    }
};

int main(){
    vector<vector<int>> friends;
    vector<int> child;
    int rows = 0, t = 0;
    cout << "number of friends: ";
    cin >> rows;
    for (int i = 1; i <= rows; i++){
        for (int j = 1; j <= rows; j++){
            cin >> t;
            child.push_back(t);
        }
        friends.push_back(child);
        child.clear();
    }
    Solution sol;
    cout << sol.findCircleNum2(friends);
    return 0;
}
/*
有 n 个城市，其中一些彼此相连，另一些没有相连。如果城市 a 与城市 b 直接相连，且城市 b 与城市 c 直接相连，那么城市 a 与城市 c 间接相连。

省份 是一组直接或间接相连的城市，组内不含其他没有相连的城市。

给你一个 n x n 的矩阵 isConnected ，其中 isConnected[i][j] = 1 表示第 i 个城市和第 j 个城市直接相连，而 isConnected[i][j] = 0 表示二者不直接相连。

返回矩阵中 省份 的数量。

 

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/number-of-provinces
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。*/
// number of friends: 3
// 1 1 0
// 1 1 0
// 0 0 1
// 2

File: LeetCode547.java
import java.util.Scanner;
import java.util.Queue;
import java.util.LinkedList;

public class LeetCode547{
    public static void main(String[] args) {
        LeetCode547 solution = new LeetCode547();
        Scanner scanner = new Scanner(System.in);
        int [][] grid;
        int rows = 0, columns = 0;
        System.out.print("number of friends: ");
        rows = scanner.nextInt();
        grid = new int[rows][rows];
        for (int i = 1; i <= rows; i++){
            for (int j = 1; j <= rows; j++){
                grid[i-1][j-1] = scanner.nextInt();
            }
        }
        System.out.println(solution.findCircleNum(grid));
        scanner.close();
    }

    public int findCircleNum(int[][] isConnected) {
        int provinces = isConnected.length;
        boolean[] visited = new boolean[provinces];
        int circles = 0;
        Queue<Integer> queue = new LinkedList<Integer>();
        for (int i = 0; i < provinces; i++) {
            if (!visited[i]) {
                queue.offer(i);
                while (!queue.isEmpty()) {
                    int j = queue.poll();
                    visited[j] = true;
                    for (int k = 0; k < provinces; k++) {
                        if (isConnected[j][k] == 1 && !visited[k]) {
                            queue.offer(k);
                        }
                    }
                }
                circles++;
            }
        }
        return circles;
    }
}
// number of friends: 3
// 1 1 0
// 1 1 0
// 0 0 1
// 2